htlcswitch/link_test.go

```diff
@@ -0,0 +1,626 @@
+// mode by setting the 'Curve" parameter to nil. Doing this avoids printing out
+// each of the field elements in the curve parameters for secp256k1.
+func messageToString(msg lnwire.Message) string {
+	switch m := msg.(type) {
+	case *lnwire.RevokeAndAck:
+		m.NextRevocationKey.Curve = nil
+	case *lnwire.NodeAnnouncement:
+		m.NodeID.Curve = nil
+	case *lnwire.ChannelAnnouncement:
+		m.NodeID1.Curve = nil
+		m.NodeID2.Curve = nil
+		m.BitcoinKey1.Curve = nil
+		m.BitcoinKey2.Curve = nil
+	case *lnwire.SingleFundingComplete:
+		m.RevocationKey.Curve = nil
+	case *lnwire.SingleFundingRequest:
+		m.CommitmentKey.Curve = nil
+		m.ChannelDerivationPoint.Curve = nil
+	case *lnwire.SingleFundingResponse:
+		m.ChannelDerivationPoint.Curve = nil
+		m.CommitmentKey.Curve = nil
+		m.RevocationKey.Curve = nil
+	}
+
+	return spew.Sdump(msg)
+}
+
+// createLogFunc is a helper function which returns the function which will be
+// used for logging message are received from another peer.
+func createLogFunc(name string, channelID lnwire.ChannelID) messageInterceptor {
+	return func(m lnwire.Message) {
+		if getChanID(m) == channelID {
+			// Skip logging of extend revocation window messages.
+			switch m := m.(type) {
+			case *lnwire.RevokeAndAck:
+				var zeroHash chainhash.Hash
+				if bytes.Equal(zeroHash[:], m.Revocation[:]) {
+					return
+				}
+			}
+
+			fmt.Printf("---------------------- \n %v received: "+
+				"%v", name, messageToString(m))
+		}
+	}
+}
+
+// TestChannelLinkSingleHopPayment in this test we checks the interaction
+// between Alice and Bob within scope of one channel.
+func TestChannelLinkSingleHopPayment(t *testing.T) {
+	n := newThreeHopNetwork(t)
+	if err := n.start(); err != nil {
+		t.Fatal(err)
+	}
+	defer n.stop()
+
+	bobBandwidthBefore := n.firstBobChannelLink.Bandwidth()
+	aliceBandwidthBefore := n.aliceChannelLink.Bandwidth()
+
+	debug := false
+	if debug {
+		// Log message that alice receives.
+		n.aliceServer.record(createLogFunc("alice",
+			n.aliceChannelLink.ChanID()))
+
+		// Log message that bob receives.
+		n.bobServer.record(createLogFunc("bob",
+			n.firstBobChannelLink.ChanID()))
+	}
+
+	var amount btcutil.Amount = btcutil.SatoshiPerBitcoin
+	errChan, invoice, err := n.makePayment(n.bobServer, amount)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for:
+	// * HTLC add request to be sent to bob.
+	// * alice<->bob commitment state to be updated.
+	// * settle request to be sent back from bob to alice.
+	// * alice<->bob commitment state to be updated.
+	// * user notification to be sent.
+	select {
+	case err := <-errChan:
+		if err != nil {
+			t.Fatalf("something wrong went when sending request: "+
+				"%v", err)
+		}
+		break
+	case <-time.After(time.Second):
+		t.Fatal("htlc was no settled in time")
+	}
+
+	// Wait for Bob to receive the revocation.
+	time.Sleep(100 * time.Millisecond)
+
+	// Check that alice invoice was settled and bandwidth of HTLC
+	// links was changed.
+	if !invoice.Terms.Settled {
```

Comment:
This seems to rely on the fact that this mock implementation directly modifies the invoice struct via a pointer, rather than through a database layer?